Fix/mobile sidebar revamp

nikrb opened the pull request:
supercedes pr 132 fix/mobile-sidebar.

Major changes to sidebar so we don't lose tree state after hiding the material-ui drawer (mobile only).
Slimmed down CMSContainer component.
Html element details/summary didn't work as a controlled component so replace with home grown version.
Move Collapsible test to utils (as that's where they belong) and create a new Collapsible specific snapshot test.
Create tests for CMSContainer and add tests to Sidebar and Sidebar utils.

nikrb committed:
fix: mobile Sidebar

make sure cms tree is built on mobile load
update cms sidebar article links to use material-ui/Drawer
 for selected highlight
make sidebar child of Drawer div to dismiss sidebar onclick

nikrb committed:
fix: keep sidebar open until article clicked

nikrb committed:
refactor: Sidebar as functional component

nikrb committed:
test: export getTree and getChildren for test

nikrb committed:
make onArticleSelect required prop

nikrb committed:
fix: Collapsible unit test

nikrb committed:
rewrite sidebar tree handling to prepare for mobile fixes

nikrb committed:
fix: linting

nikrb committed:
fix: sidebar tree state for mobile

implement collabpsible without html details/summary element
stop expand events bubbling up tree
build the tree html when sidbar drawer is opened for mobile

nikrb committed:
fix: tree default property expanded

nikrb committed:
update styling

nikrb committed:
update test suite

moved collapsible tests into utils and create collapsible test
create tests for expandTree functionality

nikrb committed:
fix: prettier code formatting

nikrb committed:
fix: right arrow code for sidebar tree

nikrb committed:
test: update collapsible snapshot

nikrb committed:
remove old commented code

nikrb committed:
add Sidebar tests

nikrb committed:
test: add tests to Sidebar utils

nikrb committed:
test: add test to Collapsible

nikrb committed:
test: refactor CMSContainer and create test

jenovs commented on frontend/src/Sidebar/Collapsible.test.js:
```diff
@@ -1,94 +1,31 @@
+      {{ _id: '1', title: 'Some article title' }}
+    </Collapsible>
+  );
+  comp.instance().onClick({ stopPropagation: () => {}});
```
`Prettier` re-formats this line.

jenovs commented on frontend/src/Sidebar/Sidebar.js:
```diff
@@ -1,51 +1,141 @@
+      mobile = true;
+    }
+    return mobile;
+  };
```
Could be refactored out of component and get called with parameters:
```js
const checkMobile = (mobile, windowWidth) => {
  const BREAK_MOBILE = 900;

  if (
    (mobile && windowWidth > BREAK_MOBILE) ||
    (!mobile && windowWidth <= BREAK_MOBILE)
  ) {
    return !mobile;
  }

  return mobile;
};
```

jenovs commented on frontend/src/Sidebar/Sidebar.test.js:
```diff
@@ -0,0 +1,142 @@
+  params: { id: '' },
+};
+
+function flushPromises() {
```
Linter is not happy :unamused:

jenovs commented on frontend/src/Sidebar/Sidebar.test.js:
```diff
@@ -0,0 +1,142 @@
+    expect(inst.state.open).toBe(true);
+    expect(comp).toMatchSnapshot();
+  });
+});
```
Passing of this test depends on the previous test. Not good.
Edit: actually looks like all of the test are depending on each other :unamused:

nikrb replied:
It seemed natural for one test to run on the results of the last - I take it this is bad practice. Will fix

nikrb replied:
fixed with beforeEach but doesn't seem to have been caught (this review still shows for me)

jenovs replied:
EDIT: I don't think it's natural for one test to run on the results of the last.
Better way to accomplish that would be:
```js
it('should test some stuff', () => {
  // initial setup
  // assert that setup is correct
  // simulate user interaction, e.g. click
  // assert changes
  // simulate another interaction
  // assert changes
  // etc.
}
```

nikrb replied:
@jenovs is this comment out of order? you're saying tests following on isn't natural ... or do you mean the beforeEach isn't the best way to fix?

nikrb replied:
oh I see you mean put them in the same test - ok. Will merge shortly

jenovs replied:
Looks like we wrote our comments almost simultaneously :grinning: 
I edited my comment.
EDIT: Looks like GitHub comments are not quite real time.

jenovs commented on frontend/src/Sidebar/utils.test.js:
```diff
@@ -0,0 +1,78 @@
+
+it('should create expanded tree if id present', () => {
+  const articleTree = getTree(articles);
+  const html = buildHtml(articles, articleTree, '101', onArticleSelect, onExpand);
```
`Prettier` re-formats this line.

nikrb committed:
fix: review changes

nikrb committed:
fix: move onClick handler to Summary

trigger onClick when summary is clicked, not parent as article clicks were causing
detail parent to be closed

nikrb committed:
fix: linting

nikrb committed:
test: update snapshot
